core/src/main/java/tc/oc/pgm/spawner/Spawner.java

```diff
@@ -0,0 +1,148 @@
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.entity.EntityDeathEvent;
+import org.bukkit.event.entity.ItemDespawnEvent;
+import org.bukkit.event.player.PlayerPickupItemEvent;
+import org.bukkit.event.player.PlayerQuitEvent;
+import org.bukkit.util.Vector;
+import tc.oc.pgm.api.PGM;
+import tc.oc.pgm.api.event.CoarsePlayerMoveEvent;
+import tc.oc.pgm.api.feature.Feature;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.Tickable;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.api.time.Tick;
+import tc.oc.pgm.util.TimeUtils;
+import tc.oc.pgm.util.bukkit.OnlinePlayerMapAdapter;
+
+public class Spawner implements Feature<SpawnerDefinition>, Listener, Tickable {
+
+  private final SpawnerDefinition definition;
+  private final Match match;
+
+  private long lastTick;
+
+  private int spawnedEntities;
+  private OnlinePlayerMapAdapter<MatchPlayer> trackedPlayers;
+
+  public static final String METADATA_KEY = "PGM_SPAWNER_OBJECT";
+
+  private static long generatedDelay;
+
+  public Spawner(SpawnerDefinition definition, Match match) {
+    this.definition = definition;
+    this.match = match;
+
+    this.lastTick = match.getTick().tick;
+    this.trackedPlayers = new OnlinePlayerMapAdapter<>(PGM.get());
+    generateDelay();
+  }
+
+  @Override
+  public String getId() {
+    return definition.id;
+  }
+
+  @Override
+  public SpawnerDefinition getDefinition() {
+    return definition;
+  }
+
+  @Override
+  public void tick(Match match, Tick tick) {
+    if (!canSpawn()) {
+      return;
+    }
+    if (match.getTick().tick - lastTick >= generatedDelay) {
+      for (Spawnable spawnable : definition.objects) {
+        Vector randomSpawnLoc = definition.spawnRegion.getRandom(match.getRandom());
+        spawnable.spawn(
+            new Location(
+                match.getWorld(),
+                randomSpawnLoc.getX(),
+                randomSpawnLoc.getY(),
+                randomSpawnLoc.getZ()),
+            match);
+        if (spawnable.isTracked()) {
+          spawnedEntities = spawnedEntities + spawnable.getSpawnCount();
+        }
+      }
+      generateDelay();
+      lastTick = match.getTick().tick;
+    }
+  }
+
+  private void generateDelay() {
+    if (definition.minDelay == definition.maxDelay) {
+      generatedDelay = TimeUtils.toTicks(definition.delay);
+    } else {
+      long maxDelay = TimeUtils.toTicks(definition.maxDelay);
+      long minDelay = TimeUtils.toTicks(definition.minDelay);
+      generatedDelay =
+          (long)
+              (match.getRandom().nextDouble() * (maxDelay - minDelay)
+                  + minDelay); // Picks a random tick duration between minDelay and maxDelay
+    }
+  }
+
+  private boolean canSpawn() {
+    if (spawnedEntities >= definition.maxEntities || trackedPlayers.isEmpty()) {
+      return false;
+    }
```

Comment:
Can be one line
`if (spawnedEntities >= definition.maxEntities || trackedPlayers.isEmpty()) return false;`